core/src/main/java/org/apache/druid/timeline/VersionedIntervalTimeline.java

```diff
@@ -753,6 +757,67 @@ private void remove(
     return retVal;
   }
 
+  @VisibleForTesting
+  public static List<String> getUniqueDimensions(
```

Comment:
The place of these methods seems a bit unintuitive to me. Their first parameter is a list of `TimelineObjectHolder`s because I think it's easier to use in `DruidInputSource` and `IngestSegmentFirehose`. But, in fact, a list of `DataSegments` would be enough. Is there any other place that needs to use these methods other than `DruidInputSource` or `IngestSegmentFirehose`? If not, I guess a new util class such as `CompactionUtils` or `DataSegments` might be better?

Reply:
I moved these to a new class `ReingestionTimelineUtils` and added some javadocs

Reply:
The existing comments within the method mention ordering based on recency of segments, so I kept the input as TimelineObjectHolders

Reply:
Hm, so I had suggested moving them to `VersionedIntervalTimeline` because they seem like generic timeline operations, not specific to re-ingesting segments. Is having a `ReingestionTimelineUtils` better than their original home if they are still going to live somewhere specific to re-ingesting segments?

Reply:
Hmm, my only view on this is it should be somewhere other than `IngestSegmentFirehoseFactory` since that's headed for eventual deprecation. 

It could go in `VersionedIntervalTimeline` although that class is a bit weird (it accepts generic types but also has methods that specifically use DataSegment). I feel like the utils class is fine for now

Reply:
My stance is similar to @jon-wei. It could be used for more general use cases, but I don't see them now. Maybe we can move them from `ReingestionTimelineUtils` to somewhere else if we need in the future.